### PR TITLE
Automated cherry pick of #101737: Use CSI driver to determine unique name for migrated in-tree

### PR DESCRIPTION
**Pull request by ialidzhikov**
Cherry pick of #101737 on release-1.18.

#101737: Use CSI driver to determine unique name for migrated in-tree

For details on the cherry pick process, see the [cherry pick requests](https://git.k8s.io/community/contributors/devel/sig-release/cherry-picks.md) page.

### CODE REVIEW

**Comment by ialidzhikov** [comment]
/area storage
/kind bug

This PR is a fix for a regression that was introduced in 1.18.16 with https://github.com/kubernetes/kubernetes/pull/98224.
/kind regression

/cc @Jiawei0227 @msau42

PS. I really know that 1.18 is out of support and there will be an exceptional last patch release for 1.18 (https://groups.google.com/g/kubernetes-dev/c/KuF8s2zueFs/m/GWGAACgKAgAJ). But I think the bug fix #101737 deserves a cherry-pick.

**Comment by puerco** [comment]
Hi @ialidzhikov I'm sorry, 1.18 is now EOL so we cannot approve the cherry-pick anymore :(

**Comment by ialidzhikov** [comment]
> Hi @ialidzhikov I'm sorry, 1.18 is now EOL so we cannot approve the cherry-pick anymore :(

According to https://groups.google.com/g/kubernetes-dev/c/KuF8s2zueFs/m/GWGAACgKAgAJ it seems that there will be one final patch release. I know that 1.18 is already out of support but I believe that this fix is critical enough to be included as well. The regression itself was introduced in 1.18.16. @Jiawei0227 @msau42 I believe you would agree with me.

File: pkg/controller/volume/attachdetach/attach_detach_controller_test.go [hunk 1]
```diff
@@ -393,7 +415,17 @@ func Test_ADC_VolumeAttachmentRecovery(t *testing.T) {
 func volumeAttachmentRecoveryTestCase(t *testing.T, tc vaTest) {
 	fakeKubeClient := controllervolumetesting.CreateTestClient()
 	informerFactory := informers.NewSharedInformerFactory(fakeKubeClient, time.Second*1)
-	plugins := controllervolumetesting.CreateTestPlugin()
+	var plugins []volume.VolumePlugin
+	if tc.csiMigration {
+		defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.CSIMigration, tc.csiMigration)()
+		defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.CSIMigrationGCE, tc.csiMigration)()
+		defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.InTreePluginGCEUnregister, tc.csiMigration)()
```
Comment (Jiawei0227) [inline pkg/controller/volume/attachdetach/attach_detach_controller_test.go:422]: InTreePluginGCEUnregister does not exist in 1.18. You may want to change it to CSIMigrationGCEComplete.

**Review by Jiawei0227 — COMMENTED** [review]
There is another PR that is related to this which is: https://github.com/kubernetes/kubernetes/pull/102176/files

I believe if you really want to backport it in 1.18 we should also cherrypick this.

**Comment by ialidzhikov** [comment]
> There is another PR that is related to this which is: https://github.com/kubernetes/kubernetes/pull/102176/files
> 
> I believe if you really want to backport it in 1.18 we should also cherrypick this.

I synced with the PR author @timebertt few days back on this topic. We agreed that the change is not that critical because it only fixes the garbage collection logic for VolumeAttachments to consider CSI migrated volumes as well. It is nice to have of course but I don't want to bundle it into this PR.
On the other side #101737 fixes an issue where kube-controller-manager can potentially wrongly detach CSI migrated volumes. This issue is already reported in https://github.com/kubernetes/cloud-provider-openstack/issues/1531, https://github.com/kubernetes-sigs/aws-ebs-csi-driver/issues/896 and we see it on our landscapes as well.

**Comment by ialidzhikov** [comment]
All checks are passing now. @Jiawei0227 @msau42 @gnufied can you check again?

PS: I guess the final patch release for 1.18 is scheduled for 2021-06-16.

**Comment by msau42** [comment]
/approve

This fixes a regression that causes disks using CSI migration to be prematurely detached from nodes while pods are still running, causing IO errors to workloads if kube-controller-manger restarts (like on an upgrade). I think this is a serious enough bug to be considered for the final patch release.

**Comment by justaugustus** [comment]
Slack thread: https://kubernetes.slack.com/archives/CJH2GBF7Y/p1623453145304800

**Pull request closed without merging**